### PR TITLE
Ported to ESP32

### PR DESCRIPTION
**Pull request by scruzin**

### CODE REVIEW

File: arduino/netsender/NetSender.cpp [hunk 1]
```diff
@@ -25,20 +25,43 @@
 #include <limits.h>
 #include <string.h>
 #include <ctype.h>
+
+#define ESP32
```
Comment (saxon-milton) [inline arduino/netsender/NetSender.cpp:29]: I think there is no need to manually set this depending on the platform; ```ESP32``` and ```ESP8266``` should be predefined macros from what I've read online, and then re linux, you should be able to check for ```__linux__``` assuming you're using GCC.

File: arduino/netsender/NetSender.cpp [hunk 2]
```diff
@@ -25,20 +25,43 @@
...
 #include <ESP8266WiFi.h>
 #include <ESP8266HTTPClient.h>
 #include <WiFiClient.h>
 #include <EEPROM.h>
-//#include "nonarduino.h"
+#endif
+#ifdef ESP32
+#include <WiFi.h>
+#include <HTTPClient.h>
+#include <EEPROM.h>
```
Comment (saxon-milton) [inline arduino/netsender/NetSender.cpp:40]: This is probably something that could happen later, but apparently ```EEPROM.h``` is deprecated for the ESP32 with a preference for ```preferences.h``` which has the following advantages 

1. No need to call ```EEPROM.commit();``` data is saved automatically.
2. Key-value based storage.
3. Has wear levelling to reduce wear and tear on flash.

File: arduino/netsender/NetSender.cpp [hunk 3]
```diff
@@ -25,20 +25,43 @@
...
+
+#define ESP32
+
+#ifdef ESP8266
 #include <ESP8266WiFi.h>
 #include <ESP8266HTTPClient.h>
 #include <WiFiClient.h>
 #include <EEPROM.h>
-//#include "nonarduino.h"
+#endif
```
Comment (saxon-milton) [inline arduino/netsender/NetSender.cpp:36]: Consider using ```#elif``` instead

File: arduino/netsender/NetSender.cpp [hunk 4]
```diff
@@ -25,20 +25,43 @@
...
 
 #include "NetSender.h"
 
 namespace NetSender {
 
 // Hardware constants.
-#define ALARM_PIN              0    // GPIO pin indicating an alarm. Also controls the red LED.
-#define LED_PIN                2    // GPIO pin corresponding to blue LED.
+#ifdef ESP8266
+#define ALARM_PIN              0    // GPIO pin indicating an alarm. Controls the red LED.
```
Comment (saxon-milton) [inline arduino/netsender/NetSender.cpp:52]: Consider using ```constexpr``` for all these consts, that way you can have type safety.

File: arduino/netsender/NetSender.cpp [hunk 5]
```diff
@@ -262,57 +309,67 @@ int setPins(const char * names, Pin * pins) {
...
+    level = PowerPins[ii].on ? HIGH : LOW;
+    if (alarm) {
+      level = LOW;
     }
     pinMode(PowerPins[ii].pin, OUTPUT);
-    digitalWrite(PowerPins[ii].pin, LOW);
-    if (Debug) Serial.print(F("Reset power pin: D")), Serial.println(PowerPins[ii].pin);
-  }
+    digitalWrite(PowerPins[ii].pin, level);
+    if (Debug) Serial.print(F("Set power pin: D")), Serial.print(PowerPins[ii].pin), Serial.print(F(" ")), Serial.println(fmtLevel(level));
```
Comment (saxon-milton) [inline arduino/netsender/NetSender.cpp:324]: I think there is ```Serial.printf``` available. Also, could we have a ```debug``` function that checks ```Debug``` ?

File: arduino/netsender/NetSender.cpp [hunk 6]
```diff
@@ -700,19 +753,20 @@ bool httpRequest(String url, String body, String& reply) {
     url = http.header("Location");
     if (Debug) Serial.print(F("Redirecting to: ")), Serial.println(url);
     http.end();
+    client.stop();
     return httpRequest(url, body, reply); // Redirect to the new location.
   }
 
-  if (status == httpOK) {
+  bool ok = (status == httpOK);
```
Comment (saxon-milton) [inline arduino/netsender/NetSender.cpp:760]: consider using ```auto```

File: arduino/netsender/NetSender.cpp [hunk 7]
```diff
@@ -796,7 +850,9 @@ bool request(RequestType req, Pin * inputs, Pin * outputs, bool * reconfig, Stri
   }
 
   if (extractJson(reply, "rc", param)) {
-    switch (param.toInt()) {
+    int rc = param.toInt();
```
Comment (saxon-milton) [inline arduino/netsender/NetSender.cpp:853]: Consider using ```auto```

**Review by saxon-milton — APPROVED** [review]
LGTM besides some small things!

File: arduino/netsender/NetSender.cpp (hunk 1 above)
Comment (scruzin) [reply arduino/netsender/NetSender.cpp:29]: I think you're right. And, yes, I use gcc when compiling without Arduino.

File: arduino/netsender/NetSender.cpp (hunk 2 above)
Comment (scruzin) [reply arduino/netsender/NetSender.cpp:40]: Yes, a separate PR.

File: arduino/netsender/NetSender.cpp (hunk 5 above)
Comment (scruzin) [reply arduino/netsender/NetSender.cpp:322]: I didn't realize that ESP supported Serial.printf as the standard Arduino library does not . The origins of this code goes back to the Arduino Uno and predates the ESP8266 and ESP32!

I'll make a separate PR to tidy up debug logging.

File: arduino/netsender/NetSender.cpp (hunk 6 above)
Comment (scruzin) [reply arduino/netsender/NetSender.cpp:758]: Yup.

File: arduino/netsender/NetSender.cpp (hunk 3 above)
Comment (scruzin) [reply arduino/netsender/NetSender.cpp:34]: I prefer separate #ifdef for each hardware type. Personal preference, I suppose.

**Pull request merged**